resources/channel.sub.js

```diff
@@ -0,0 +1,1003 @@
+            return "remoteobject";
+        }
+        if (obj instanceof SendChannel) {
+            return "sendchannel";
+        }
+        if (obj instanceof RecvChannel) {
+            return "recvchannel";
+        }
+        if (obj instanceof Error) {
+            return "error";
+        }
+        if (Array.isArray(obj)) {
+            return "array";
+        }
+        let constructor = obj.constructor && obj.constructor.name;
+        if (constructor === "RegExp" ||
+            constructor === "Date" ||
+            constructor === "Map" ||
+            constructor === "Set" ||
+            constructor == "WeakMap" ||
+            constructor == "WeakSet") {
+            return constructor.toLowerCase();
+        }
+        // The handling of cross-global objects here is broken
+        if (typeof window == "object" && window === self) {
+            if (obj instanceof Element) {
+                return "element";
+            }
+            if (obj instanceof Document) {
+                return "document";
+            }
+            if (obj instanceof Node) {
+                return "node";
+            }
+            if (obj instanceof Window) {
+                return "window";
+            }
+        }
+        if (Promise.resolve(obj) === obj) {
+            return "promise";
+        }
+        return "object";
+    }
+
+    let remoteObjectsById = new Map();
+
+    function remoteId(obj) {
+        let rv;
+        rv = createUuid();
+        remoteObjectsById.set(rv, obj);
+        return rv;
+    }
+
+    /**
+     * Representation of a non-primitive type passed through a channel
+     */
+    class RemoteObject {
+        constructor(type, objectId) {
+            this.type = type;
+            this.objectId = objectId;
+        }
+
+        /**
+         * Create a RemoteObject containing a handle to reference obj
+         *
+         * @param {Any} obj - The object to reference.
+         */
+        static from(obj) {
+            let type = typeName(obj);
+            let id = remoteId(obj);
+            return new RemoteObject(type, id);
+        }
+
+        /**
+         * Return the local object referenced by the ``objectId`` of
+         * this ``RemoteObject``, or ``null`` if there isn't a such an
+         * object in this realm.
+         */
+        toLocal() {
+            if (remoteObjectsById.has(this.objectId)) {
+                return remoteObjectsById.get(this.objectId);
+            }
+            return null;
+        }
+
+        /**
+         * Remove the object from the local cache. This means that future
+         * calls to ``toLocal`` with the same objectId will always return
+         * ``null``.
+         */
+        delete() {
+            if (remoteObjectsById.has(this.objectId)) {
+                remoteObjectsById.delete(this.objectId);
+            }
+        }
+    }
+
+    self.RemoteObject = RemoteObject;
+
+    function serialize(obj) {
+        const stack = [{item: obj}];
```

Comment:
We use `stack.push(...)` and stack.shift()`, so it's a queue, not a stack:

```suggestion
        const queue = [{item: obj}];
```

A comment describing what the items in the queue are and what the possible properties are would be really helpful too. I think that if I hadn't seen traversal-by-stack (or queue) a lot of times and hadn't reviewed the WebDriver BiDi serialization format, I'd be 100% lost here.